### DISCUSSION TITLE
Edit Keybinds of a Transient State

### DISCUSSION THREAD

**Discussion by eeriemyxi** [syl20bnr/spacemacs#16675] [Q&A] [answered]
How to remap the keybinds of a transient state? I use Colemak keyboard layout and I have been wanting to edit the very QWERTY-biased shortcuts of various transient states like `spacemacs/scroll-transient-state` for example. I searched everywhere and couldn't find anything anywhere about it.

Since I couldn't find any documentation on remapping them I also tried making my own custom layer that implements a transient layer in a fashion that built-in layers of Spacemacs define; the idea was that I would replace the biased transient with my own. The relevant documentation at https://develop.spacemacs.org/doc/LAYERS.html speaks of some "Spacemacs' layer search path" but I can't find a semblance of documentation about it. I'm too clueless as a beginner to be welcomed to these documentation pages I guess.

**Comment by fnussbaum** [comment, marked as answer]
You can remove and add/remap bindings to existing transient states in your `dotspacemacs/user-config` with the functions `spacemacs/transient-state-register-remove-bindings` and `spacemacs/transient-state-register-add-bindings`. For example, in my configuration I used to have:
```elisp
  (spacemacs/transient-state-register-remove-bindings 'layouts
    '("w"))
  (spacemacs/transient-state-register-add-bindings 'layouts
    '(("k" spacemacs/workspaces-transient-state/body :exit t)))
```

For private layers: Spacemacs looks for them in the directories specified by the variable `configuration-layer-private-layer-directory` and the user option `dotspacemacs-configuration-layer-path`. You can use <kbd>SPC SPC configuration-layer/create-layer</kbd> to automatically create a private layer in one of those directories.
(Redefining transient states in your dotfile should also work, without having to create private layers.)

By the way, are you using the keyboard-layout layer? Perhaps this layer could be extended to automatically remap some of those transient state bindings, too.

**Reply by eeriemyxi** [reply]
Thank you for the in-depth explanation, I am sure it will help others out. I hope search engines soon pick it up. Personally I ended up having to switch to my own configuration which I very much wanted to avoid because it takes quite a lot of maintenance. Especially when you use something other than QWERTY.

> By the way, are you using the keyboard-layout layer?

Yes I was using it. I was using `colemak-hnei` to be specific.

I originally chose Spacemacs due its official support for other keyboard layouts but I was let down by the fact that it doesn't actually apply the changes everywhere relevant, but only in some places. There were other issues, like [this one](https://github.com/syl20bnr/spacemacs/issues/8729#issuecomment-2478765656) which I never managed to get to the bottom of, and then there was an issue with `mc` configuration layer (with the default `evil-mc` backend) where if you undo after using it, it will corrupt your undo history and also delete a huge chunk of your buffer. I thought it was spacemacs-specific but having used it in my own config, I later figured out the issue was with the undo backend. Eventually I ended getting rid of anything `evil` related entirely and switching to Meow. There were issues with `php`'s configuration layer too but I don't remember the details anymore.

**Reply by fnussbaum** [reply]
Thanks for your reply!

> I originally chose Spacemacs due its official support for other keyboard layouts but I was let down by the fact that it doesn't actually apply the changes everywhere relevant, but only in some places.

I think remapping transient state bindings sounds like a reasonable extension to the keyboard-layout layer, and should not take too much work. I might look into it. In any case bug reports, feature requests, and pull requests are always welcome.

> where if you undo after using it, it will corrupt your undo history

I assume this happened with the `undo-tree` backend? I am aware that undo-tree does not support the undo elements created by the `combine-change-calls` macro in full generality, often leading to corrupted histories when a command uses it. This bug is on my list to report upstream, though the package does not seem to be maintained anymore. Maybe we should discourage the use of undo-tree even more explicitly in the meantime.

> There were issues with php's configuration layer too

I can imagine that some of the lesser used layers deserve more maintenance. Any help would be appreciated.
